--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -48,10 +_,15 @@
         if (playerx instanceof ServerPlayer serverplayer) {
             Container container = this.container;
             if (container instanceof AbstractFurnaceBlockEntity abstractfurnaceblockentity) {
+                // Ketting start
+                abstractfurnaceblockentity.ketting$awardUsedRecipesAndPopExperience$itemstack.set(p_39558_);
+                abstractfurnaceblockentity.ketting$awardUsedRecipesAndPopExperience$amount.set(this.removeCount);
+                // Ketting end
                 abstractfurnaceblockentity.awardUsedRecipesAndPopExperience(serverplayer);
             }
         }
 
         this.removeCount = 0;
+        net.neoforged.neoforge.event.EventHooks.firePlayerSmeltedEvent(this.player, p_39558_);
     }
 }
