--- a/net/minecraft/world/entity/vehicle/ChestBoat.java
+++ b/net/minecraft/world/entity/vehicle/ChestBoat.java
@@ -212,4 +_,51 @@
     public void stopOpen(Player p_270286_) {
         this.level().gameEvent(GameEvent.CONTAINER_CLOSE, this.position(), GameEvent.Context.of(p_270286_));
     }
+
+    // CraftBukkit start
+    public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<org.bukkit.entity.HumanEntity>();
+    private int maxStack = MAX_STACK;
+
+    @Override
+    public java.util.List<ItemStack> getContents() {
+        return this.itemStacks;
+    }
+
+    @Override
+    public void onOpen(org.bukkit.craftbukkit.v1_20_R3.entity.CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    @Override
+    public void onClose(org.bukkit.craftbukkit.v1_20_R3.entity.CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    @Override
+    public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    @Override
+    public org.bukkit.inventory.InventoryHolder getOwner() {
+        org.bukkit.entity.Entity entity = getBukkitEntity();
+        if (entity instanceof org.bukkit.inventory.InventoryHolder) return (org.bukkit.inventory.InventoryHolder) entity;
+        return null;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return maxStack;
+    }
+
+    @Override
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+
+    @Override
+    public org.bukkit.Location getLocation() {
+        return getBukkitEntity().getLocation();
+    }
+    // CraftBukkit end
 }
