--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -63,6 +_,8 @@
                         }
                     }
 
+                    mob.ketting$setTarget$reason.set(org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+                    mob.ketting$setTarget$fireEvent.set(true);
                     this.mob.setTarget(livingentity);
                     return true;
                 }
@@ -83,6 +_,8 @@
 
     @Override
     public void stop() {
+        mob.ketting$setTarget$reason.set(org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET);
+        mob.ketting$setTarget$fireEvent.set(true);
         this.mob.setTarget(null);
         this.targetMob = null;
     }
