--- a/net/minecraft/world/damagesource/DamageSource.java
+++ b/net/minecraft/world/damagesource/DamageSource.java
@@ -19,7 +_,38 @@
     private final Entity directEntity;
     @Nullable
     private final Vec3 damageSourcePosition;
-
+    // CraftBukkit start
+    private boolean sweep;
+    private boolean melting;
+    private boolean poison;
+
+    public boolean isSweep() {
+        return sweep;
+    }
+
+    public DamageSource sweep() {
+        this.sweep = true;
+        return this;
+    }
+
+    public boolean isMelting() {
+        return melting;
+    }
+
+    public DamageSource melting() {
+        this.melting = true;
+        return this;
+    }
+
+    public boolean isPoison() {
+        return poison;
+    }
+
+    public DamageSource poison() {
+        this.poison = true;
+        return this;
+    }
+    // CraftBukkit end
     @Override
     public String toString() {
         return "DamageSource (" + this.type().msgId() + ")";
@@ -88,6 +_,10 @@
         return this.type().msgId();
     }
 
+    /**
+     * @deprecated Use {@link DamageScaling#getScalingFunction()}
+     */
+    @Deprecated(since = "1.20.1")
     public boolean scalesWithDifficulty() {
         return switch(this.type().scaling()) {
             case NEVER -> false;
