--- a/net/minecraft/world/effect/HealOrHarmMobEffect.java
+++ b/net/minecraft/world/effect/HealOrHarmMobEffect.java
@@ -16,7 +_,7 @@
     public void applyEffectTick(LivingEntity p_295255_, int p_295147_) {
         super.applyEffectTick(p_295255_, p_295147_);
         if (this.isHarm == p_295255_.isInvertedHealAndHarm()) {
-            p_295255_.heal((float)Math.max(4 << p_295147_, 0));
+            p_295255_.heal((float)Math.max(4 << p_295147_, 0), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
             p_295255_.hurt(p_295255_.damageSources().magic(), (float)(6 << p_295147_));
         }
@@ -26,7 +_,7 @@
     public void applyInstantenousEffect(@Nullable Entity p_294574_, @Nullable Entity p_295692_, LivingEntity p_296483_, int p_296095_, double p_295178_) {
         if (this.isHarm == p_296483_.isInvertedHealAndHarm()) {
             int i = (int)(p_295178_ * (double)(4 << p_296095_) + 0.5);
-            p_296483_.heal((float)i);
+            p_296483_.heal((float)i, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
             int j = (int)(p_295178_ * (double)(6 << p_296095_) + 0.5);
             if (p_294574_ == null) {
